### PR TITLE
[core] Ant Task Formatter encoding issue with XMLRenderer

### PR DESCRIPTION
**Pull request by pyxide**
Hi,

The writers in the Ant task's formatter use the default platform's encoding (for instance cp1252 on Western Windows) and the XMLRenderer sets the default XML encoding to UTF-8. This may lead to incorrect XML encoding, detected for instance when the XML file is sent to a XSLT task.
This goes undetected as long as the characters in the XML file are in the ASCII subset, which is frequently the case with code, usually in plain Latin language. In my case, the issue was raised because of the localized formatting of some line numbers greater than 999, due to the localized thousand separator 0xA0, which is out of the ASCII subset.

I modified the Formatter class to always use a charset for writers. When no encoding is specified, UTF-8 is used instead of the default platform's encoding.

I strongly recommend to *always set the encoding parameter* of the formatter element. It should be reflected as so in the Ant Task documentation.

```xml
<pmd encoding="cp1252"> <!-- source encoding -->
  <formatter type="xml" toFile="pmd.xml">
    <param name="encoding" value="UTF-8" /> <!-- report encoding -->
  </formatter>
  <fileset dir="src/main/java">
    <include name="**/*.java"/>
  </fileset>
</pmd>
```
Note: when digging into sourceforge issues for similar bugs, I found [Bug 877: Output with encoding CP1252](https://sourceforge.net/p/pmd/bugs/877/) . Not the same task, but the same symptoms, probably the same cause ?

### CODE REVIEW

**Comment by pyxide** [comment]
The build fails because I use `IOUtils.closeQuietly(Closeable...)` which does not exist in commons-io-2.4, but is present in commons-io-2.5. If there is no plan to upgrade this library, it should read `IOUtils.closeQuietly(input);IOUtils.closeQuietly(writer);`

File: pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java [hunk 1]
```diff
@@ -57,12 +59,27 @@ public Renderer getRenderer() {
...
 
     public void start(String baseDir) {
+
+        Properties properties = createProperties();
+        
+        Charset charset;
+        {
+            String s = (String) properties.get("encoding");
+            if (null == s) {
+                charset = Charset.forName("UTF-8");
```
Comment (pyxide) [inline pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java]: It's done, plus two distinct calls to IOUtils.closeQuietly

**Comment by jsotuyod** [comment]
Great, thanks.

Please, check that checkstyle has failed on Travis.

**Comment by jsotuyod** [comment]
@pyxide thanks a lot for your contribution. I'll take some time to look into this, ideally I would like to be able to reproduce the issue in a unit test. Once I'm able to verify the issue and this fix, I'll merge it and update documentation as needed.

I should be able to get it merged for 5.4.4, 5.5.3 and later.

**Comment by jsotuyod** [comment]
@pyxide I'm having trouble trying to reproduce the issue.

[Line numbers are not localized by the XML renderer](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/renderers/XMLRenderer.java#L91-L92) when generating output. `StringBuilder.append()` simply appends `Integer.getChars()` on the given number, which does no localization (at least on my JVM).

Could you provide a complete sample for the error including:
 - [ ] Default charset / locale used when running ant
 - [ ] Ant build configuration
 - [ ] JVM version
 - [ ] Sample code
 - [ ] Output PMD report demonstrating the issue

Thanks in advanced

**Comment by pyxide** [comment]
Hi @jsotuyod,

here is another simple way to reproduce the issue, even without Ant.

1. Force the file system encoding for Java to Cp1252 in the launcher script (here pmd.bat)
```batch
@echo off
set TOPDIR=%~dp0..
set OPTS=-Dfile.encoding=Cp1252
set MAIN_CLASS=net.sourceforge.pmd.PMD

java -classpath %TOPDIR%\lib\* %OPTS% %MAIN_CLASS% %*
```
2. Create a simple faulty Java class `UnusedVariableWithNonAsciiChar.java`, encoded as UTF-8 (notice the e acute and the a grave).
```java
public class UnusedVariableWithNonAsciiChar {
	private String déjàVu;
}
```
3. Run this command (`-encoding utf-8` specifies the Java source code encoding)
```batch
pmd.bat -d UnusedVariableWithNonAsciiChar.java -f xml -r pmd-report.xml -R java-unusedcode -language java -encoding utf-8
```
The report file `pmd-report.xml` should be incorrect (open it in Firefox or any other strict XML reader, it will raise an error), because of the e acute and a grave are encoded by the writer with the Cp1252 charset when the XML encoding says UTF-8.
If you set `set OPTS=-Dfile.encoding=UTF-8` and run this test again, it should be OK, because the writer encoding matches the encoding declared in the XML file.

The two generated PMD reports:
- [pmd-report.when-file-encoding-utf-8.xml.txt](https://github.com/pmd/pmd/files/683246/pmd-report.when-file-encoding-utf-8.xml.txt)
- [pmd-report.when-file-encoding-cp1252.malformed.xml.txt](https://github.com/pmd/pmd/files/683245/pmd-report.when-file-encoding-cp1252.malformed.xml.txt)

I'm working on the simple reproducible sample with Ant.

**Comment by pyxide** [comment]
Here is a sample of the localisation effect on encoding I observed.

`pmd.bat` with different options
```batch
@echo off
set TOPDIR=%~dp0..

@rem OK : ',' thousand separator in ASCII subset, common to UTF-8 and Cp1252, the XML is well-formed.
@rem set OPTS=-Duser.language=en -Duser.country=US -Dfile.encoding=Cp1252
@rem set OPTS=-Duser.language=en -Duser.country=US -Dfile.encoding=us-ascii

@rem OK : the writer is set to UTF-8 as the declared XML encoding.
@rem set OPTS=-Duser.language=fr -Duser.country=FR -Dfile.encoding=UTF-8

@rem FAIL : u00a0 thousand separator, not part of ASCII subset, the XML is malformed.
@rem u00a0 is encoded as one byte 0xA0 by Cp1252, instead of the two expected UTF-8 bytes 0xC2 0xA0 
@rem set OPTS=-Duser.language=fr -Duser.country=FR -Dfile.encoding=Cp1252

set MAIN_CLASS=net.sourceforge.pmd.PMD

java -classpath %TOPDIR%\lib\* %OPTS% %MAIN_CLASS% %*
```
I put the target Java file in attachement, it has more than 1000 lines
[MoreThanThousandLinesOfCodeWithDuplicateLitterals.java.txt](https://github.com/pmd/pmd/files/683582/MoreThanThousandLinesOfCodeWithDuplicateLitterals.java.txt)

Then run this script, while successively testing the OPTS variables in `pmd.bat`
```batch
pmd.bat -d org/example/pmd/MoreThanThousandLinesOfCodeWithDuplicateLitterals.java -f xml -r pmd-report.xml -R java-strings -language java -encoding utf-8
```

**Comment by adangel** [comment]
Hi @pyxide 
I can confirm the bug and as you already figured out, it's not only happening, when running PMD via ant. The renderers (e.g. XMLRenderer) are provided with a ready to use `Writer` - and when the Writers are created, the default (platform) charset is used. This is done both when the formatter for the Ant Task creates the writer and when PMD itself creates the writer for the renderer.
Although the XMLRenderer has the encoding property, this is only used as-is for the encoding xml attribute and doesn't actually change the encoding.
The workaround is of course, to always use UTF-8/Unicode - you'll be avoiding lots of trouble. You can do this (as you already know) by modifying `pmd.bat` and setting the option like this:

`set OPTS=-Dfile.encoding=UTF-8`

Note: the system property `file.encoding` should be sufficient.

As for the thousand separators - I see it now. Like the variable name, it is in the message of the violations and not in the XML attribute values, e.g.

```
<violation beginline="1023" endline="1023" begincolumn="31" endcolumn="51" rule="AvoidDuplicateLiterals" ruleset="String and StringBuffer" package="org.example.pmd" class="MoreThanThousandLinesOfCodeWithDuplicateLitterals" externalInfoUrl="https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/strings.html#AvoidDuplicateLiterals" priority="3">
The String literal &quot;duplicated litteral&quot; appears 4 times in this file; the first occurrence is on line 1.023
</violation>
```

Btw. `U+00A0` (encoded in UTF-8 as 0xC2 0xA0) is a no-break space character. Since I used a German Locale, my thousand separator is a dot, French seems to use the no-break space, in English, it should be a comma.

When we format the message, we use `java.text.MessageFormat.format(...)`, which obviously outputs the numbers in localized format. [AvoidDuplicateLiteralsRule message](https://github.com/pmd/pmd/blob/master/pmd-java/src/main/resources/rulesets/java/strings.xml#L13) and [AbstractRuleViolationFactory](https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRuleViolationFactory.java#L19-L34).

In summary: main issue is, that it's not possible to actually change the encoding of the renderers, when PMD is writing to a file (with an without ant).

File: pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java [hunk 2]
```diff
@@ -57,12 +61,26 @@ public Renderer getRenderer() {
...
+                properties.put("encoding", charset.name());
+            } else {
+                charset = Charset.forName(s);
+            }
+        }
+
         try {
             if (toConsole) {
-                writer = new BufferedWriter(new OutputStreamWriter(System.out));
+                writer = new BufferedWriter(new OutputStreamWriter(System.out, charset));
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java]: Should the console should always use the platform encoding? I would expect this, as otherwise you might see strange characters. However, if you redirect the output to a file, then we should use the specified encoding...

See also http://stackoverflow.com/questions/1403772/how-can-i-check-if-a-java-programs-input-output-streams-are-connected-to-a-term - we could use [System.console()](http://docs.oracle.com/javase/6/docs/api/java/lang/System.html#console()) for a simple check.

**Comment by jsotuyod** [comment]
@pyxide thanks for your feedback! This is definitely an issue. Unfortunately writing a unit test for this is non trivial due to depending on properties used by Java to execute.

Non the less, the fix can be tested and is certainly working. I believe that with the one improvement detected by @adangel, we could add the new documentation and get this merged.

File: pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java (hunk 2 above)
Comment (pyxide) [inline pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java]: > Should the console should always use the platform encoding?

On every platform but Windows I would say yes. On Windows it's trickier because the platform's encoding is different from the console's encoding. For instance on my Windows, the platform's encoding is Cp1252 (`file.encoding` property) and the console's encoding is IBM850 (`sun.stderr.encoding` property).
So, if you want to view directly the output in the console, use IBM850, but if you redirect the output to a file, you may want to use Cp1252 or UTF-8 for instance.

So, again, it boils down to *always set the encoding parameter*, depending on your needs.

**Comment by pyxide** [comment]
@jsotuyod 
> None the less, the fix can be tested and is certainly working

yes, I reckon it's only a little fix for the XML renderer in Ant and does not address the whole issue, but it's a start. I'll take a look at the rest and may come up with new patchs.

@adangel 

> Although the XMLRenderer has the encoding property, this is only used as-is for the encoding xml attribute and doesn't actually change the encoding.

It's nevertheless mandatory that the XML encoding attribute matches the writer's encoding. 

> As for the thousand separators - I see it now. Like the variable name, it is in the message of the violations and not in the XML attribute values, e.g.

It could also be in an attribute, in a path for instance
`<file name="C:\déjàVu\pmd\src\UnusedVariableWithNonAsciiChar.java">`

> Btw. U+00A0 (encoded in UTF-8 as 0xC2 0xA0) is a no-break space character. Since I used a German Locale, my thousand separator is a dot, French seems to use the no-break space, in English, it should be a comma.

Yes, i wrote a note in that sense to explain the goal of this test file [of this test file MoreThanThousandLinesOfCodeWithDuplicateLitterals.java.txt](https://github.com/pmd/pmd/files/683582/MoreThanThousandLinesOfCodeWithDuplicateLitterals.java.txt)

File: pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java (hunk 2 above)
Comment (jsotuyod) [inline pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java]: I'm curious here, but if we obtained the writer to a connected console through [`Console.writer()`](https://docs.oracle.com/javase/7/docs/api/java/io/Console.html#writer()), wouldn't that writer have the correct charset already configured?

Comment (pyxide) [inline pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java]: Yes, but I was not sure that we could get the encoding buried in the writer. We need it to set the encoding attribute in the XML file. However, using the system properties `file.encoding` and `sun.stdout.encoding`, it's possible.

Comment (jsotuyod) [inline pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java]: But sun.* properties are non standard and may not be present on different JREs...

I'll think about this on depth over the weekend. Thanks again for your efforts.

**Comment by pyxide** [comment]
Here is the [Ant test file](https://github.com/pmd/pmd/files/691211/build-pmd-test.xml.txt), self-contained, two properties to be set (the PMD lib home and a test folder)

I updated my patch and now that the encoding is really passed to the renderer, the characters which are not in the ASCII subset are escaped with numeric entities if the encoding is not UTF-8, and the whole XML content basically becomes ASCII compatible. Then it was hard to tell if the writer's encoding was valid with no testing content. So, I tweaked the XMLRenderer (forced `useUTF8 = true` in the `start()` method) to write characters instead of characters and ran a test (see at the end).

There are many formatters in the build file, some to test the console output, some to test the file output.

The XML file reports' well-formedness is tested with a simple XSLT task.

### Interactive Console ###
The first formatter uses by default IBM850, which is the valid console's encoding on my Windows.
The second formatter's encoding is forced to IBM850.
```ant
P:\dev\ws\workspace\make>ant -f build-pmd-test.xml
Buildfile: P:\dev\ws\workspace\make\build-pmd-test.xml

properties:

prepare:

create-source:

pmd-test:
      [pmd] <?xml version="1.0" encoding="IBM850"?>
      [pmd] <pmd version="5.5.2" timestamp="2017-01-07T02:41:36.586">
      [pmd] <file name="P:\dev\test\pmd\src\UnusedVariableWithNonAsciiChar.java">
      [pmd] <violation beginline="2" endline="2" begincolumn="24" endcolumn="29" rule="UnusedPrivateField" ruleset="Unused Code" class="UnusedVariableWithNonAsciiChar" variable="d&#xe9;j&#xe0;Vu" externalInfoUrl="https://pmd.github.io/pmd-5.5.2/pmd-java/rules/java/unusedcode.html#UnusedPrivateField" priority="3">
      [pmd] Avoid unused private fields such as 'd&#xe9;j&#xe0;Vu'.
      [pmd] </violation>
      [pmd] </file>
      [pmd] <?xml version="1.0" encoding="IBM850"?>
      [pmd] <pmd version="5.5.2" timestamp="2017-01-07T02:41:36.586">
      [pmd] <file name="P:\dev\test\pmd\src\UnusedVariableWithNonAsciiChar.java">
      [pmd] <violation beginline="2" endline="2" begincolumn="24" endcolumn="29" rule="UnusedPrivateField" ruleset="Unused Code" class="UnusedVariableWithNonAsciiChar" variable="d&#xe9;j&#xe0;Vu" externalInfoUrl="https://pmd.github.io/pmd-5.5.2/pmd-java/rules/java/unusedcode.html#UnusedPrivateField" priority="3">
      [pmd] Avoid unused private fields such as 'd&#xe9;j&#xe0;Vu'.
      [pmd] </violation>
      [pmd] </file>
      [pmd] </pmd>
      [pmd] </pmd>
     [xslt] Processing P:\dev\test\pmd\report\pmd-report.cp1252.xml to P:\dev\test\pmd\report\pmd-report.cp1252.xml.styled.xml
     [xslt] Loading stylesheet P:\dev\test\pmd\src\identity.xsl
     [xslt] Processing P:\dev\test\pmd\report\pmd-report.utf-8.xml to P:\dev\test\pmd\report\pmd-report.utf-8.xml.styled.xml
     [xslt] Loading stylesheet P:\dev\test\pmd\src\identity.xsl

BUILD SUCCESSFUL
Total time: 1 second
```
### Piped or Redirected Console ###

The first formatter uses by default Cp1252, which is the valid platform's encoding on my Windows.
The second formatter's encoding is set to IBM850.

```ant
P:\dev\ws\workspace\make>echo.|ant -f build-pmd-test.xml
Buildfile: P:\dev\ws\workspace\make\build-pmd-test.xml

properties:

prepare:

create-source:

pmd-test:
      [pmd] <?xml version="1.0" encoding="windows-1252"?>
      [pmd] <pmd version="5.5.2" timestamp="2017-01-07T02:41:52.911">
      [pmd] <file name="P:\dev\test\pmd\src\UnusedVariableWithNonAsciiChar.java">
      [pmd] <violation beginline="2" endline="2" begincolumn="24" endcolumn="29" rule="UnusedPrivateField" ruleset="Unused Code" class="UnusedVariableWithNonAsciiChar" variable="d&#xe9;j&#xe0;Vu" externalInfoUrl="https://pmd.github.io/pmd-5.5.2/pmd-java/rules/java/unusedcode.html#UnusedPrivateField" priority="3">
      [pmd] Avoid unused private fields such as 'd&#xe9;j&#xe0;Vu'.
      [pmd] </violation>
      [pmd] </file>
      [pmd] <?xml version="1.0" encoding="IBM850"?>
      [pmd] <pmd version="5.5.2" timestamp="2017-01-07T02:41:52.911">
      [pmd] <file name="P:\dev\test\pmd\src\UnusedVariableWithNonAsciiChar.java">
      [pmd] <violation beginline="2" endline="2" begincolumn="24" endcolumn="29" rule="UnusedPrivateField" ruleset="Unused Code" class="UnusedVariableWithNonAsciiChar" variable="d&#xe9;j&#xe0;Vu" externalInfoUrl="https://pmd.github.io/pmd-5.5.2/pmd-java/rules/java/unusedcode.html#UnusedPrivateField" priority="3">
      [pmd] Avoid unused private fields such as 'd&#xe9;j&#xe0;Vu'.
      [pmd] </violation>
      [pmd] </file>
      [pmd] </pmd>
      [pmd] </pmd>
     [xslt] Processing P:\dev\test\pmd\report\pmd-report.cp1252.xml to P:\dev\test\pmd\report\pmd-report.cp1252.xml.styled.xml
     [xslt] Loading stylesheet P:\dev\test\pmd\src\identity.xsl
     [xslt] Processing P:\dev\test\pmd\report\pmd-report.utf-8.xml to P:\dev\test\pmd\report\pmd-report.utf-8.xml.styled.xml
     [xslt] Loading stylesheet P:\dev\test\pmd\src\identity.xsl

BUILD SUCCESSFUL
Total time: 1 second
```
### Piped or Redirected Console, Tweaked XMLRenderer without Numeric Entities  ###

As expected, the first console formatter encodes with Cp1252, and some gibberish `dÚjÓVu` is displayed in the IBM850 console.
The second formatter's encoding is set to IBM850, and outputs the correct bytes that read `déjàVu`.

```ant
P:\dev\ws\workspace\make>echo.|ant -f build-pmd-test.xml
Buildfile: P:\dev\ws\workspace\make\build-pmd-test.xml

properties:

prepare:

create-source:

pmd-test:
      [pmd] <?xml version="1.0" encoding="windows-1252"?>
      [pmd] <pmd version="5.5.2" timestamp="2017-01-07T03:49:36.263">
      [pmd] <file name="P:\dev\test\pmd\src\UnusedVariableWithNonAsciiChar.java">
      [pmd] <violation beginline="2" endline="2" begincolumn="24" endcolumn="29" rule="UnusedPrivateField" ruleset="Unused Code" class="UnusedVariableWithNonAsciiChar" variable="dÚjÓVu" externalInfoUrl="https://pmd.github.io/pmd-5.5.2/pmd-java/rules/java/unusedcode.html#UnusedPrivateField" priority="3">
      [pmd] Avoid unused private fields such as 'dÚjÓVu'.
      [pmd] </violation>
      [pmd] </file>
      [pmd] <?xml version="1.0" encoding="IBM850"?>
      [pmd] <pmd version="5.5.2" timestamp="2017-01-07T03:49:36.263">
      [pmd] <file name="P:\dev\test\pmd\src\UnusedVariableWithNonAsciiChar.java">
      [pmd] <violation beginline="2" endline="2" begincolumn="24" endcolumn="29" rule="UnusedPrivateField" ruleset="Unused Code" class="UnusedVariableWithNonAsciiChar" variable="déjàVu" externalInfoUrl="https://pmd.github.io/pmd-5.5.2/pmd-java/rules/java/unusedcode.html#UnusedPrivateField" priority="3">
      [pmd] Avoid unused private fields such as 'déjàVu'.
      [pmd] </violation>
      [pmd] </file>
      [pmd] </pmd>
      [pmd] </pmd>
     [xslt] Processing P:\dev\test\pmd\report\pmd-report.cp1252.xml to P:\dev\test\pmd\report\pmd-report.cp1252.xml.styled.xml
     [xslt] Loading stylesheet P:\dev\test\pmd\src\identity.xsl
     [xslt] Processing P:\dev\test\pmd\report\pmd-report.utf-8.xml to P:\dev\test\pmd\report\pmd-report.utf-8.xml.styled.xml
     [xslt] Loading stylesheet P:\dev\test\pmd\src\identity.xsl

BUILD SUCCESSFUL
Total time: 1 second
```

File: pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java (hunk 2 above)
Comment (pyxide) [inline pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java]: > But sun.* properties are non standard and may not be present on different JREs...

I know, it's only best effort. Note that it could then default to `file.encoding` instead of UTF-8, which might be more sensible for console output.

Comment (jsotuyod) [inline pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java]: The only "cross-jre" workaround I've been able to find so far is [this one](http://hg.openjdk.java.net/code-tools/jmh/file/0797222e066e/jmh-core/src/main/java/org/openjdk/jmh/util/Utils.java#l173). Being used by JMH makes me think it's pretty solid and covers pretty much all JREs. I'll keep looking into this tomorrow.

Comment (jsotuyod) [inline pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java]: I can confirm the `sun.stdout.encoding` property is present in Oracle JRE and OpenJDK, but not on other JREs such as IBM's J9. However, the reflection based approach used by JMH does work on all 3 JREs, and most probably all others, since it's based on the API exposed in class sources that are freely available.

I'd suggest we adopt the reflection detection code for console encoding from JMH for maximum compatibility and leave the rest as is.

File: pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java [hunk 3]
```diff
@@ -63,18 +67,17 @@ public Renderer getRenderer() {
...
         Charset charset;
         {
             String s = (String) properties.get("encoding");
             if (null == s) {
 
                 if (toConsole) {
-                    if (null == System.console()) {
-                        // pipe or redirect, no interactive console.
+                    s = getConsoleEncoding();
+                    if (null == s) {
```
Comment (jsotuyod) [inline pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java]: @pyxide doesn't this mean that, even if the output is piped (`System.console() == null`), we would take the value returned by `Console.encoding()`?

I would just used the reflection approach wen there is a console attached.

File: pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java (hunk 2 above)
Comment (pyxide) [inline pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java]: I added the reflection code to get the console encoding. Thank you for suggesting the console as a reflection target, it's more effective and portable than any attempt with a writer.

File: pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java (hunk 3 above)
Comment (pyxide) [inline pmd-core/src/main/java/net/sourceforge/pmd/ant/Formatter.java]: @jsotuyod yes it means that, I described the behaviour in the commit's details. I can change it or course.

**Comment by jsotuyod** [comment]
@pyxide amazing work! Thanks for bearing with us.

I'll try to get this merged as soon as I've some spare time to make it available for the next release.

**Comment by jsotuyod** [comment]
This is merged and will be included in PMD 5.4.4, 5.5.3 and 5.6.0

I'll be writing documentation shortly for the newly added property.
